iterator/cdc.go

```diff
@@ -0,0 +1,265 @@
+		lastSequenceNumber: nil,
+		streamArn:          streamArn,
+		tomb:               &tomb.Tomb{},
+		cache:              make(chan stypes.Record, 10), // todo size?
+		ticker:             time.NewTicker(pollingPeriod),
+		p:                  p,
+	}
+	shardIterator, err := c.getShardIterator(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get shard iterator: %w", err)
+	}
+	c.shardIterator = shardIterator
+
+	// start listening to changes
+	c.tomb.Go(c.startCDC)
+
+	return c, nil
+}
+
+// HasNext returns a boolean that indicates whether the iterator has any objects in the buffer or not.
+func (c *CDCIterator) HasNext(_ context.Context) bool {
+	return len(c.cache) > 0 || !c.tomb.Alive() // if tomb is dead we return true so caller will fetch error with Next
+}
+
+// Next returns the next record in the iterator.
+func (c *CDCIterator) Next(ctx context.Context) (opencdc.Record, error) {
+	var rec stypes.Record
+	select {
+	case r := <-c.cache:
+		rec = r
+	case <-c.tomb.Dead():
+		return opencdc.Record{}, fmt.Errorf("tomb is dead: %w", c.tomb.Err())
+	case <-ctx.Done():
+		return opencdc.Record{}, ctx.Err()
+	}
+
+	return c.getOpenCDCRec(rec)
+}
+
+func (c *CDCIterator) Stop() {
+	// stop the goRoutine
+	c.ticker.Stop()
+	c.tomb.Kill(errors.New("cdc iterator is stopped"))
+}
+
+// startCDC gets records from the stream using the shardIterators,
+// then adds the records to a buffered cache.
+func (c *CDCIterator) startCDC() error {
+	defer close(c.cache)
+
+	for {
+		select {
+		case <-c.tomb.Dying():
+			return fmt.Errorf("tomb is dying: %w", c.tomb.Err())
+		case <-c.ticker.C: // detect changes every polling period
+			out, err := c.streamsClient.GetRecords(c.tomb.Context(nil), &dynamodbstreams.GetRecordsInput{ //nolint:staticcheck // SA1012 tomb expects nil
+				ShardIterator: c.shardIterator,
+			})
+			if err != nil {
+				return fmt.Errorf("error getting records: %w", err)
+			}
+
+			// process the stream records
+			for _, record := range out.Records {
+				select {
+				case c.cache <- record:
+					// Successfully sent record
+				case <-c.tomb.Dying():
+					return fmt.Errorf("tomb is dying: %w", c.tomb.Err())
+				}
+			}
+
+			// update the shard iterator
+			c.shardIterator = out.NextShardIterator
+
+			if c.shardIterator == nil {
+				c.shardIterator, err = c.getShardIterator(c.tomb.Context(nil)) //nolint:staticcheck // SA1012 tomb expects nil
+				if err != nil {
+					return fmt.Errorf("failed to get shard iterator: %w", err)
+				}
+			}
+		}
+	}
+}
+
+// getShardIterator gets the shard iterator for the stream.
+func (c *CDCIterator) getShardIterator(ctx context.Context) (*string, error) {
+	// Describe the stream to get the shard ID
+	describeStreamOutput, err := c.streamsClient.DescribeStream(ctx, &dynamodbstreams.DescribeStreamInput{
+		StreamArn: aws.String(c.streamArn),
+	})
+	if err != nil {
+		return nil, fmt.Errorf("failed to describe stream: %w", err)
+	}
+
+	// Start from the last shard
```

Comment:
Is there a special reason for that choice?

Reply:
the other shards will contain older events, the user could have created the stream long before they used conduit, and we just want to get the latest data, which will be added to the last shard.

Reply:
You said "and we just want to get the latest data". How can we know what's the latest data, that the connector should read?

Reply:
I meant the latest data from which we ran the connector, cuz that version didn't support continuing CDC, but now this is all changed :D